[Providers/HTTP] Add adapter parameter to HttpHook to allow custom requests adapters

jieyao-MilestoneHub opened the pull request:
This PR adds an `adapter` parameter to the `HttpHook`, allowing users to mount custom adapters for HTTP request handling. This enhances flexibility and supports use cases like custom retries, timeouts, and SSL handling.

### Summary of Changes:
- Added `adapter` parameter to `HttpHook` to allow custom HTTP adapters.
- Modified `get_conn` to support mounting custom adapters or using TCPKeepAliveAdapter by default.
- Added comprehensive tests to validate the functionality of the `adapter` parameter and its integration with `get_conn`.
- Ensured all new tests pass and maintain compatibility with existing functionality.

### Issue Reference:
Closes #44285

### Testing and Validation:
- Added unit tests in `~/providers/tests/http/hooks/test_http.py` to validate:
  - Mounting custom adapters.
  - Default behavior when no adapter is provided.
  - Handling invalid adapter types.
- Ran `pytest providers/tests/http/hooks/test_http.py` using breeze.
- Pass `pre-commit run --files providers/src/airflow/providers/http/hooks/http.py providers/tests/http/hooks/test_http.py` using breeze.

### Backward Compatibility:
This change is backward-compatible. The new `adapter` parameter is optional and defaults to None, preserving existing behavior.


<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Commit:
feat(http-hook): add adapter parameter to HttpHook and enhance get_conn

- Added `adapter` parameter to `HttpHook` to allow custom HTTP adapters.
- Modified `get_conn` to support mounting custom adapters or using TCPKeepAliveAdapter by default.
- Added comprehensive tests to validate the functionality of the `adapter` parameter and its integration with `get_conn`.
- Ensured all new tests pass and maintain compatibility with existing functionality.

shahar1 commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -72,6 +74,7 @@ def __init__(
         method: str = "POST",
         http_conn_id: str = default_conn_name,
         auth_type: Any = None,
+        adapter: BaseAdapter | None = None,
```
Please add an appropriate docstring in the section above

shahar1 commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -143,6 +151,18 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
+                interval=self.keep_alive_interval,
+            )
+            session.mount("http://", keep_alive_adapter)
+            session.mount("https://", keep_alive_adapter)
```
We do we need to instantiate `TCPKeepAliveAdapter` here if we already have it in the `run` method?

shahar1 reviewed (commented):
Welcome to Apache Airflow and thanks for your contribution!
Looks OK overall, I have some comments :)

Commit:
fix(http_hook): Update docstring and remove redundant TCPKeepAliveAdapter

- Added missing `adapter` parameter description to the HttpHook class docstring.
- Removed redundant instantiation of `TCPKeepAliveAdapter` in the `run` method since it's already instantiated in `get_conn`.

Commit:
fix(http_hook): improve get_conn session setup and TCP adapter logic

- Ensured proper mounting of TCP Keep-Alive adapter when enabled.
- Improved handling of connection extras for cleaner session configuration.

Commit:
feat(http): update get_conn logic and corresponding tests (#44302)

Aligned the `get_conn` method with the adjustments specified in #44302,
including refined handling of headers. Optimized and updated test cases
to ensure compatibility and maintain robust test coverage.

shahar1 reviewed (approved):
Looks good to me :)
I'll be happy for an additional review though

ashb commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -83,11 +87,25 @@ def __init__(
+                interval=self.keep_alive_interval,
+            )
+        else:
+            self.keep_alive_adapter = None
```
What's this all about?

ashb replied:
Ah just moved. 

This doesn't make any network requests by creating it here does it?

kiaradlf replied:
it doesn't, adapters just modify how actual requests are handled

ashb commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -102,49 +120,64 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
+            session.mount(f"{scheme}://", self.adapter)
+        elif self.keep_alive_adapter:
+            session.mount("http://", self.keep_alive_adapter)
+            session.mount("https://", self.keep_alive_adapter)
```
Why do we parse the scheme in one branch but hardcode http+https in the other?

jieyao-MilestoneHub replied:
Thank you for your insight. The reasoning behind this implementation is to account for custom cases where the schema might be empty. However, this does seem to tie back to _set_base_url. I’ll make adjustments accordingly. Appreciate you pointing this out!

Commit:
refactor(http_hook): simplify HttpHook by reverting BaseAdapter to HTTPAdapter

- Changed the `adapter` parameter to accept only `HTTPAdapter` instead of `BaseAdapter`.
- Strengthened `_set_base_url` validation to ensure base_url is constructed with stricter conditions.
- Adjusted `_mount_adapters` to improve maintainability.

Commit:
refactor(http_hook): simplify HttpHook by reverting BaseAdapter to HTTPAdapter

- Changed the `adapter` parameter to accept only `HTTPAdapter` instead of `BaseAdapter`.
- Strengthened `_set_base_url` validation to ensure base_url is constructed with stricter conditions.
- Adjusted `_mount_adapters` to improve maintainability.

Commit:
Merge branch 'main' into feature/http-hook-custom-adapter

Commit:
Merge branch 'main' into feature/http-hook-custom-adapter

Commit:
Merge: new main

Commit:
Merge remote-tracking branch 'upstream/main'

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -72,6 +75,7 @@ def __init__(
         method: str = "POST",
         http_conn_id: str = default_conn_name,
         auth_type: Any = None,
+        adapter: HTTPAdapter | None = None,
```
We should move this one to the last parameter or make it a keyword only argument. Otherwise, it would be a breaking change. I would suggest making it a keyword only argument

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -102,49 +120,68 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
 
         return session
 
+    def _set_base_url(self, connection) -> None:
```
```suggestion
    def _set_base_url(self, connection: Connection) -> None:
```

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -102,49 +120,68 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
+        else:
+            self.base_url = f"{schema}://{host}" if host else f"{schema}://"
+            if connection.port:
+                self.base_url += f":{connection.port}"
```
```suggestion
                self.base_url = f"{self.base_url}:{connection.port}"
```

Lee-W replied:
and this

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -102,49 +120,68 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
+        if not parsed.scheme:
+            raise ValueError(f"Invalid base URL: Missing scheme in {self.base_url}")
+
+    def _set_auth(self, session: requests.Session, connection) -> None:
```
```suggestion
    def _set_auth(self, session: requests.Session, connection: Connection) -> None:
```

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -102,49 +120,68 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
+        elif self._auth_type:
+            session.auth = self.auth_type()
+
+    def _set_extra(self, session: requests.Session, connection) -> None:
```
```suggestion
    def _set_extra(self, session: requests.Session, connection: Connection) -> None:
```

Lee-W replied:
`_set_extra` is a confusing name in this case as we're not actually setting extra but extracting attributes from extra

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -102,49 +120,68 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
+            session.auth = self.auth_type()
+
+    def _set_extra(self, session: requests.Session, connection) -> None:
+        if connection.extra:
```
We should probably do this `if` check in `get_conn` instead

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -102,49 +120,68 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
+        else:
+            self.base_url = f"{schema}://{host}" if host else f"{schema}://"
+            if connection.port:
+                self.base_url += f":{connection.port}"
```
```suggestion
        if connection.port:
            self.base_url += f":{connection.port}"
```

I think it should be dedented 🤔  Could you please check whether I'm mistaken? If I'm not, would be better if we could add one unit test to cover it

jieyao-MilestoneHub replied:
If :// is present in the host string, it indicates the presence of a scheme component, and the string can be considered a complete URL as defined by [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986.html#section-3).

Lee-W replied:
got it. sounds good! Thanks for the explanation!

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -102,49 +120,68 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
+        if connection.login:
+            session.auth = self.auth_type(connection.login, connection.password)
+        elif self._auth_type:
+            session.auth = self.auth_type()
```
Setting `session` this way would work. But it might confuse some contributors as we're changing the argument. I might suggest reason the `session` object and reassign it in `get_conn` instead.

something like `session = self._set_auth(...)`

jieyao-MilestoneHub replied:
The main reason I wrote it this way was for consistency, as the nearby functions also make modifications internally.

jieyao-MilestoneHub replied:
Could you share your thoughts on the reasoning outlined above?

Lee-W replied:
Yes, the main reason is that it's not as intuitive. It's much easier to understand if it's change a value in instance level (i.e., `self. = ...`) or returning a value and assigning it back to something

jieyao-MilestoneHub replied:
Got it, thank you for your suggestion! I've made the changes accordingly.

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -102,49 +120,68 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
+        if connection.login:
+            session.auth = self.auth_type(connection.login, connection.password)
+        elif self._auth_type:
+            session.auth = self.auth_type()
```
```suggestion
    def _extract_auth(self, connection: Connection) -> Any | None:
        if connection.login:
            return self.auth_type(connection.login, connection.password)
        elif self._auth_type:
            return self.auth_type()
```

or this might also work

if

jieyao-MilestoneHub replied:
Based on the principle of not exposing session members in get_conn, we might add an additional function like:

```python
def _configure_session_from_auth(self, session: requests.Session, connection: Connection) -> None:
    session.auth = self._extract_auth(connection) 
    return session
```
However, I eventually decided to stick with the original version. Is there any issue with this approach?

Lee-W replied:
The main concern is https://github.com/apache/airflow/pull/44302#discussion_r1860249178

Lee-W replied:
The example you provide actually looks good to me

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -102,49 +120,68 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
+    def _mount_adapters(self, session: requests.Session) -> None:
+        scheme = urlparse(self.base_url).scheme
+        if not scheme:
+            raise ValueError("Cannot mount adapters: base_url must include a valid scheme (http or https).")
```
```suggestion
            raise ValueError(f"Cannot mount adapters: "{self.base_url}" does not include a valid scheme (http or https).")
```

Lee-W replied:
would be nice if we could also address this 🙂

Lee-W reviewed (changes requested):
I believe the keyword-only section is what we need to change the most. For the other parts, we can discuss how we want to proceed. 🙂

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
refactor: improve function naming and add type annotations

- Changed the function prefix from `_set` to `_configure_session_from` to enhance readability and better reflect its purpose.
- Added static type annotations for input parameters and return values.
- Included comments to document the design rationale following coding standards.
- Improved error message: replaced generic text with detailed and actionable messages.

Commit:
Merge branch 'main' into feature/http-hook-custom-adapter

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
fix: simplify the change of session

- Added a variable `session` after the change of session member

Commit:
Merge branch 'main' into feature/http-hook-custom-adapter

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -76,18 +79,33 @@ def __init__(
         self._retry_obj: Callable[..., Any]
         self._auth_type: Any = auth_type
+
+        if adapter is not None and not isinstance(adapter, HTTPAdapter):
```
```suggestion
        if not isinstance(adapter, HTTPAdapter):
```

Lee-W reviewed (commented):
Overall looks good. Once the remaining nitpicks and CI failure are resolved, I think we can merge it

Commit:
fix: Adjust response format.

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge branch 'main' into feature/http-hook-custom-adapter

Commit:
fix: simplify the logic

Commit:
fix(hook): ensure default HTTPAdapter in HttpHook init

The `adapter` parameter in `HttpHook` was previously required to be explicitly
set to an instance of `HTTPAdapter`. This commit modifies the `__init__`
method to assign a default `HTTPAdapter` when no adapter is provided.

Changes:
- Removed type checks for `adapter`, as default initialization guarantees correctness.
- Improved code readability and reduced potential runtime errors.

No functional changes beyond defaulting `adapter` to `HTTPAdapter`.

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge branch 'main' into feature/http-hook-custom-adapter

Commit:
feat(http_hook): add support for custom adapter in initialization

Refactored `HttpHook` to support a custom `HTTPAdapter` through the `adapter` parameter. If no adapter is provided, it defaults to `TCPKeepAliveAdapter` when `tcp_keep_alive=True`.

Test: Added `test_custom_adapter` to verify correct adapter mounting.

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge branch 'main' into feature/http-hook-custom-adapter

Commit:
fix: CI image checks / Static checks

- Adjust the length of each line of code.

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge branch 'main' into feature/http-hook-custom-adapter

Commit:
fix: Adjust indent style

- modify `assert instance` by PEP8

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge branch 'main' into feature/http-hook-custom-adapter

Commit:
fix: ruff error about `from requests.adapters import HTTPAdapter`

Commit:
Merge remote-tracking branch 'upstream/main'

Commit:
Merge branch 'main' into feature/http-hook-custom-adapter

Lee-W commented on providers/src/airflow/providers/http/hooks/http.py:
```diff
@@ -102,47 +113,76 @@ def get_conn(self, headers: dict[Any, Any] | None = None) -> requests.Session:
+    def _set_base_url(self, connection: Connection) -> None:
+        host = connection.host or ""
+        schema = connection.schema or "http"
+        # RFC 3986 (https://www.rfc-editor.org/rfc/rfc3986.html#page-16)
```
This is nice!
